chain/jsonrpc/tests/rpc_query.rs

```diff
@@ -443,6 +445,283 @@ fn test_health_ok() {
     });
 }
 
+/// Retrieve genesis config via JSON RPC.
+/// WARNING: Be mindful about changing genesis structure as it is part of the public protocol!
+#[test]
+fn test_genesis_config() {
```

Comment:
This is not a nice unit test, because it hardcodes genesis, so every time we modify genesis we'll have to replicate it here. Unit tests should test some very specific logic, in this case the RPC returns the content of a genesis config, but without records. So instead of hardcoding the config here we should fetch it from whatever the test fetches it from and compare it.

Reply:
While I agree that the values should not be compared, the structure and the value types are part of our public contract, so changing it should be noticeable right away. Loading the genesis from the source will only test that we get the same source of the genesis, but does not provide any guarantees about backward compatibility, and we have quite a few tests which actually fail to catch breaking APIs because we re-use the same internal structure, so the tests silently get upgraded to the new structure and hide the fact that something got changed.

Reply:
But it doesn't belong to the RPC crate. If we want to lock Genesis Config then we should enforce it under Genesis crate somewhere.

It's unclear what this test enforces, is this immutability of Genesis? The better solution might be splitting records and Genesis Config, and making sure Genesis Config from file doesn't use defaults.

Reply:
OK, I will rewrite the test to check the structure and types only. Do you agree that the structure should be stable?

I have checked how involved the process of splitting Genesis Config and Records is, and it seems that it is not that complex. However, I am a bit unsure about the implications of keeping those structures apart. Should we pursue it? /cc @ailisp @bowenwang1996 @mikhailOK

Reply:
We are pursuing it https://github.com/nearprotocol/nearcore/pull/2126

Reply:
@bowenwang1996 As far as I can see, you split the files, but not the structures. If I understood @evgenykuzyakov correctly, he was suggesting separating the structures as well, or I misunderstood?

Reply:
Let's get more eyes on this question. @nearmax @mikhailOK @ilblackdragon @SkidanovAlex @fckt

Reply:
@evgenykuzyakov the question is not clear here.

I agree that we should have RPC output checking, but also not convinced in belongs in the unittests. I was thinking nearlib is a better place for these tests.

I don't understand how splitting structures would help - RPC needs to return all of the data one way or another and we need to test that is consistently returned and if you change it - it should be painful (because the real pain is actually upgrading 100 other dependencies by us and community).

Reply:
Well, putting it into nearlib is a struggle for two reasons:

1. nearlib is an external repository, which makes it really clunky to implement tests for something that is currently only in a PR state
2. nearlib is implemented in JS, which has a bad notion of types and strictness, so writing strict tests gets really problematic. I wish we can move to some tooling doing the strict checks
instead of writing them by hand (but I am not aware of any good tool to do that).

Reply:
I don't argue whether to move this test to nearlib or keep it there. My argument is the unit test should test behavior of the function it tests. It shouldn't rely on the data passed through the function.

The right unit test for this function should be:
```
let expected = get_genesis_config();
let actual = RPC.get_genesis_config();
asser_eq!(actual, expected);
```

This function tests the following:
```
let expected = {chain_id: "NEAR TESTNET"...}
let actual = RPC.get_genesis_config();
asser_eq!(actual, expected);
```

Now, if we change genesis in some other unrelated place. RPC unit test fails. This is unexpected behavior. To repeat, if we want to enforce immutability of genesis config, we should not do this as RPC unittest. It should be a separate PR, that makes it clear that it enforces genesis config.

Reply:
I agree with @frol that we should have tests to check API regression, however I also agree with @evgenykuzyakov that regular RPC tests should not be asserting the content like this. Here is how I suggest we split this:
1. `tests` folder contains regular RPC tests. These tests assert direct responsibility of RPC -- providing access to the internal data of the node -- and they do not assert anything else, like the content of that data. So if we had `ViewGenesisConfig` that was mapped to `GenesisConfig` inside RPC crate then we should've asserted that this mapping is done correctly, and would've done what @frol suggests -- checking structure and types. But since we pass `GenesisConfig` as is through RPC we should be asserting just that -- that the `GenesisConfig` from RPC is equal to the one where it is read from;
2. `regression_tests` folder will have regression tests for RPC API. These tests are expensive to maintain, RPC crate will have many of them covering all RPC endpoints, and these tests will assert the RPC response against a hardcoded value like in this file here. It has upside and downside.
  * The upside is that if we do a change somewhere in the remote place (e.g. introducing or removing some runtime fee) then it will immediately break the regression test and bring @frol 's attention who is our API guardian who would immediately decide whether it is breaking to Applayer;
  * The downside is many tests failures are going to be "false positive", e.g. a value of some runtime fee changing. However we cannot have anything more complex than simply asserting the exact equivalence of RPC output, because some delicate changes might actually be breaking to Applayer (e.g. maximum size of contract arguments might actually break explorer UI by just simply increasing).

I agree with @ilblackdragon that nearlib should have RPC tests, but I think the above (1) tests should be in Rust since it will require accessing the internals. The regression tests (2) however might be written through nearlib. Regarding @frol 's concern about writing cross-repo tests, we already have those. For now I however suggests we keep (2) in Rust while @damons  is figuring out what we do with cross-repo testing.

Reply:
Simplified the test to only check the essential fields.